File: binary_tree/binary_tree.py

class binary_node:

	def __init__(self, data = None):
		self.left = None
		self.right = None
		self.data = data

	def set_data(self, data):
		self.data = data

	def create_left(self, data = None):
		self.left = binary_node(data)
		return self.left

	def create_right(self, data = None):
		self.right = binary_node(data)
		return self.right

	def get_sum(self):
		sum = 0
		if not self.data is None:
			sum += self.data
		if not self.left is None:
			sum += self.left.get_sum()
		if not self.right is None:
			sum += self.right.get_sum()
		return sum


class binary_tree:

	def __init__(self):
		self.root = binary_node(None)

	def get_root(self):
		return self.root

	def get_sum(self):
		return self.root.get_sum()

def build_binary_tree(tree):
	n0 = tree.get_root()
	n0.set_data(-1)

	n1 = n0.create_left(2)
	n2 = n0.create_right(4)

	n3 = n1.create_left(5)
	n4 = n1.create_right(6)
	n5 = n2.create_left(67)
	n6 = n2.create_right(100)

def main():
	tree = binary_tree()
	build_binary_tree(tree)
	print('Sum: {0}'.format(tree.get_sum()))

main()

File: binary_tree/binary_tree.cs

namespace DataStructures
{
	abstract class NodeValue
	{
		public abstract NodeValue Add(NodeValue rhs);
		public abstract NodeValue Accumulate(NodeValue rhs);
	}
	
	class NodeValueInt : NodeValue
	{
		public int value { get; set; }
	
		public static NodeValueInt zero { get { return new NodeValueInt(0); } }
	
		public NodeValueInt() { value = 0; }
		public NodeValueInt(int value) { this.value = value; }
	
		public override NodeValue Add(NodeValue rhs)
		{
			return new NodeValueInt((rhs as NodeValueInt).value + this.value);
		}
	
		public override NodeValue Accumulate(NodeValue rhs)
		{
			this.value += (rhs as NodeValueInt).value;
			return this;
		}
	}
	
	
	class BinaryNode<T> where T : NodeValue, new()
	{
		private T nodeValue;
		private static T zeroValue;
		private BinaryNode<T> left;
		private BinaryNode<T> right;
	
		public BinaryNode(T data = null, T zeroValue = null)
		{
			nodeValue = data;
			left = null;
			right = null;
			BinaryNode<T>.zeroValue = zeroValue;
		}
	
		public void SetSatelliteData(T data)
		{
			nodeValue = data;
		}
	
		public BinaryNode<T> CreateLeft(T data = null)
		{
			left = new BinaryNode<T>(data);
			return left;
		}
	
		public BinaryNode<T> CreateRight(T data = null)
		{
			right = new BinaryNode<T>(data);
			return right;
		}
	
		public T GetSum()
		{
			T sum = new T();
			if(nodeValue != null)
				sum.Accumulate(nodeValue);
			if(left != null)
				sum.Accumulate(left.GetSum());
			if(right != null)
				sum.Accumulate(right.GetSum());
			return sum;
		}
	}
	
	class BinaryTree<T> where T : NodeValue, new()
	{
		public BinaryNode<T> root { get; private set; }
	
		public BinaryTree(T zeroValue)
		{
			root = new BinaryNode<T>(null, zeroValue);
		}
	
		public T GetSum()
		{
			return root.GetSum();
		}
	}

}